### PR TITLE
Next generation with several improvements

### PR DESCRIPTION
**Pull request by thomasttvo**
- Fixed not receiving events after JS restarts or app crash
- New method to chunk files
- iOS only: New method to directly get status of a task
- iOS only: Fixed `cancelUpload` resolving too soon and not truly after successfully cancelling tasks
- Limit concurrent uploads to `1 per host` for iOS and `1` for Android
- Android only: Fixed cancelling an upload not reporting cancellation
- Android only: Added discretionary mode that functions similarly (but not entirely equivalent) to the iOS uploader.

Additionally, 
- Added test server to test upload behaviors
- Converted everything to typescript

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,10 +2,14 @@
...
-  "main": "src/index.js",
+  "main": "lib/index.js",
+  "typings": "lib/index.d.ts",
+  "react-native": "src/index",
+  "source": "src/index",
   "scripts": {
+    "build": "tsc && tsc-alias",
     "lint": "yarn lint:ci --fix",
-    "lint:ci": "eslint src example",
+    "lint:ci": "eslint src/** example/**",
```
Comment (elliottkember) [inline package.json:12]: I'm surprised by this one, does `src example` not work to specify these dirs as arguments?}

**Review by elliottkember — DISMISSED** [review]
👍 light speed!

Question - does chunking the file double the amount of space used? A 2GB file will take up 4GB while the upload is in progress?

File: package.json (hunk 1 above)
Comment (thomasttvo) [reply package.json:12]: Yeah. I'm guessing they stopped working at some point after some upgrade and nobody updated this

**Comment by thomasttvo** [comment]
> 👍 light speed!
> 
> Question - does chunking the file double the amount of space used? A 2GB file will take up 4GB while the upload is in progress?

Unfortunately it does. We'll need to clean the chunks a bit more frequently than others.

File: package.json (hunk 1 above)
Comment (thomasttvo) [reply package.json]: I was wrong, the error with the old method was a result of some weird behavior with this version of eslint and non `.js` files. Apparently, the right way to fix it is to add `--ext js,jsx,ts,tsx`

**Review by elliottkember — APPROVED** [review]

**Pull request merged**